### PR TITLE
ENH: Add support for more placeholders in `guess_datetime_format`

### LINKED ISSUE #43901: ENH: Add support for more placeholders in `guess_datetime_format`
#### Is your feature request related to a problem?

We recently realized that `pandas.core.tools.datetimes.guess_datetime_format` outputs an incorrect format string `"Tue %d %b %Y %H:%M:%S AM"` for the datetime `"Tue 24 Aug 2021 01:30:48 AM"`, which otherwise is parseable by `dateutil.parser.parse`.

#### Describe the solution you'd like

Since `"Tue 24 Aug 2021 01:30:48 AM"` is parseable by `dateutil.parser.parse` and `guess_datetime_format` is based on that function, it seems reasonable that `guess_datetime_format` should produce a correct format string when given this example datetime.

#### API breaking implications

I believe modifying `guess_datetime_format` to support more formats placeholders should be a backwards compatible change.

#### Describe alternatives you've considered

Not sure that there are any.

#### Additional context

```python
from pandas.core.tools.datetimes import guess_datetime_format

# actual
assert guess_datetime_format("Tue 24 Aug 2021 01:30:48 AM") == "Tue %d %b %Y %H:%M:%S AM"

# expected
assert guess_datetime_format("Tue 24 Aug 2021 01:30:48 AM") == "%a %d %b %Y %H:%M:%S %p"
```

I already have a PR up for this here: #43900

### PR DESCRIPTION
**Pull request by davesque**
Add support for day of week and meridiem placeholders and any
combination of placeholders supported by `strftime` that do not
correspond to a datetime attribute.

- [x] closes #43901
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

I'm not sure if this fix actually qualifies as a bug fix and not and enhancement.  Before, `guess_datetime_format` would output a somewhat invalid format string that included literal content from the parsed datetime such as "Tue" or "AM".

### CODE REVIEW

File: doc/source/whatsnew/v1.4.0.rst [hunk 1]
```diff
@@ -129,6 +129,7 @@ Other enhancements
 - :meth:`DataFrame.__pos__`, :meth:`DataFrame.__neg__` now retain ``ExtensionDtype`` dtypes (:issue:`43883`)
 - The error raised when an optional dependency can't be imported now includes the original exception, for easier investigation (:issue:`43882`)
 - Added :meth:`.ExponentialMovingWindow.sum` (:issue:`13297`)
+- :func:`guess_datetime_format` now correctly guesses the format of datetime strings with day of week and AM/PM placeholders (:issue:`43901`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:132]: this is not public, day ``` :func:`to_datetime` ``` and indicate what this does (speed it up?). this should go under datetime section

File: pandas/_libs/tslibs/parsing.pyx [hunk 2]
```diff
@@ -967,15 +970,18 @@ def guess_datetime_format(
...
+        # The result of `.strftime("%Z")` on a dt with no timezone is ""
+        if len(parsed_formatted) == 0:
+            continue
+
+        for i, token_format in enumerate(format_guess):
+            token_filled = tokens[i].zfill(padding)
+            if token_format is None and token_filled == parsed_formatted:
+                format_guess[i] = attr_format
+                tokens[i] = token_filled
+                found_attrs.update(attrs)
```
Comment (jreback) [inline pandas/_libs/tslibs/parsing.pyx:983]: can you add a benchmark that supports this case (i think we have them for other things that guess the format)

**Review by jreback — CHANGES_REQUESTED** [review]
cc @mroeschke

File: pandas/_libs/tslibs/parsing.pyx (hunk 2 above)
Comment (davesque) [reply pandas/_libs/tslibs/parsing.pyx:983]: Sure!

**Review by mroeschke — COMMENTED** [review]
One whatsnew comment otherwise LGTM

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
@davesque sorry for the delay. Could you merge in master one more time, and then this should be good to merge.

**Review by jreback — APPROVED** [review]
could narrow the whatsnew note to say only for '%p;'

**Pull request merged**